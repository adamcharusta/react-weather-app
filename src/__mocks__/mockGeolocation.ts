import { GeolocationType } from '../types/geolocationTypes';

export const mockGeolocation: GeolocationType = {
  id: 2553604,
  name: 'White House',
  latitude: 33.58831,
  longitude: -7.61138,
  elevation: 27,
  feature_code: 'PPLA',
  country_code: 'MA',
  admin1_id: 11281879,
  admin2_id: 6547294,
  timezone: 'Africa/Casablanca',
  population: 3144909,
  country_id: 2542007,
  country: 'Morocco',
  admin1: 'Casablanca-Settat',
  admin2: 'Casablanca',
};
