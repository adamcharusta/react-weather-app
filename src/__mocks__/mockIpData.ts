import { IpResponseType } from '../types/ipTypes';

export const mockIpData: IpResponseType = {
  ip: '45.138.153.39',
  network: '45.138.152.0/23',
  version: 'IPv4',
  city: 'Gmina Piątek',
  region: 'Łódź Voivodeship',
  region_code: '10',
  country: 'PL',
  country_name: 'Poland',
  country_code: 'PL',
  country_code_iso3: 'POL',
  country_capital: 'Warsaw',
  country_tld: '.pl',
  continent_code: 'EU',
  in_eu: true,
  postal: '99-120',
  latitude: 52.0696,
  longitude: 19.4734,
  timezone: 'Europe/Warsaw',
  utc_offset: '+0100',
  country_calling_code: '+48',
  currency: 'PLN',
  currency_name: 'Zloty',
  languages: 'pl',
  country_area: 312685,
  country_population: 37978548,
  asn: 'AS50411',
  org: 'ITV Media Sp. z o.o.',
};
