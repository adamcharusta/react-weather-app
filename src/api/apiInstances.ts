import axios from 'axios';

export const geolocationApiInstance = axios.create({
  baseURL: process.env.REACT_APP_GEOCODING_API_URL,
});
export const openMeteoApiInstance = axios.create({
  baseURL: process.env.REACT_APP_OPEN_METEO_API_URL,
  params: {
    current_weather: true,
    hourly: 'temperature_2m,relativehumidity_2m,windspeed_10m',
  },
});
export const ipApiInstance = axios.create({
  baseURL: process.env.REACT_APP_IP_API_URL,
});
