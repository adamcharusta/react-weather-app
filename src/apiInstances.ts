import axios from 'axios';

export const geolocationApiInstance = axios.create({
  baseURL: process.env.REACT_APP_GEOCODING_API_URL,
});
