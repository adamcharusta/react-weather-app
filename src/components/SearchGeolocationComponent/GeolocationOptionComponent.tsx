import React from 'react';
import { GeolocationType } from '../../types/geolocationTypes';
import { ListItem, ListItemIcon, ListItemText } from '@mui/material';
import Flag from 'react-world-flags';
import AdministrationStackComponent from '../AdministrationStackComponent/AdministrationStackComponent';

interface SearchOptionComponentPropsTypes {
  geolocation: GeolocationType;
  params: React.HTMLAttributes<HTMLLIElement>;
}

const GeolocationOptionComponent = ({ geolocation, params }: SearchOptionComponentPropsTypes) => {
  return (
    <ListItem {...params} data-testid='autocomplete-option'>
      <ListItemIcon>
        <Flag code={geolocation.country_code} style={{ width: 50 }} />
      </ListItemIcon>

      <ListItemText>
        {geolocation.name}
        <AdministrationStackComponent geolocation={geolocation} />
      </ListItemText>
    </ListItem>
  );
};

export default GeolocationOptionComponent;
