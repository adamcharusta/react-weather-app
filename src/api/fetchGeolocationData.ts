import { geolocationApiInstance } from './apiInstances';
import { GeolocationResponseType } from '../types/geolocationTypes';

const fetchGeolocationData = async (name: string): Promise<GeolocationResponseType | null> => {
  try {
    const { data } = await geolocationApiInstance.get<GeolocationResponseType>('', {
      params: {
        name,
      },
    });

    return data ? data : null;
  } catch (error) {
    console.error(error);
    return null;
  }
};

export default fetchGeolocationData;
