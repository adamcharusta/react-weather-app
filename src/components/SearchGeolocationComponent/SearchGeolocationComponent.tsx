import React, { ChangeEvent, useContext, useState } from 'react';
import { Autocomplete, Card, CardContent, CircularProgress, TextField } from '@mui/material';
import { GeolocationResponseType, GeolocationType } from '../../types/geolocationTypes';
import { geolocationApiInstance } from '../../apiInstances';
import { AxiosResponse } from 'axios';
import GeolocationOptionComponent from './GeolocationOptionComponent';
import AppContext from '../../appContext';

const SearchGeolocationComponent = () => {
  const { geolocation, setGeolocation } = useContext(AppContext);
  const [options, setOptions] = useState<GeolocationType[]>([]);
  const [loading, setLoading] = useState(false);

  const handleInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    setLoading(true);

    geolocationApiInstance
      .get('', { params: { name: event.target.value } })
      .then((res: AxiosResponse<GeolocationResponseType>) => {
        if (res.data.results) {
          setOptions(res.data.results);
        }
      })
      .finally(() => setLoading(false));
  };

  const handleAutocompleteChange = (
    event: React.SyntheticEvent,
    newValue: GeolocationType | null,
  ) => {
    setGeolocation(newValue);
  };

  // TODO: Fix "MUI: The value provided to Autocomplete is invalid." warn.
  return (
    <Card>
      <CardContent>
        <Autocomplete
          disablePortal
          id='geolocation-search'
          options={options}
          sx={{ width: '100%', marginTop: 1 }}
          loading={loading}
          value={geolocation}
          getOptionLabel={(option) => option.name}
          isOptionEqualToValue={(option, value) => option.id === value.id}
          onChange={handleAutocompleteChange}
          renderInput={(params) => (
            <TextField
              {...params}
              label='Location'
              onChange={handleInputChange}
              InputProps={{
                ...params.InputProps,
                endAdornment: (
                  <React.Fragment>
                    {loading ? <CircularProgress color='inherit' size={20} /> : null}
                    {params.InputProps.endAdornment}
                  </React.Fragment>
                ),
              }}
            />
          )}
          renderOption={(params, option) => (
            <GeolocationOptionComponent key={option.id} params={params} geolocation={option} />
          )}
        />
      </CardContent>
    </Card>
  );
};

export default SearchGeolocationComponent;
