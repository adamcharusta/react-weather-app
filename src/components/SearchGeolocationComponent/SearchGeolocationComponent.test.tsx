import { act, fireEvent, render, screen } from '@testing-library/react';
import SearchGeolocationComponent from './SearchGeolocationComponent';
import React from 'react';
import AppContext from '../../appContext';
import { mockGeolocation } from '../../__mocks__/mockGeolocation';
import MockAdapter from 'axios-mock-adapter';
import { geolocationApiInstance } from '../../api/apiInstances';

describe('SearchGeolocationComponent', () => {
  const expectedResults = [
    { ...mockGeolocation, id: 1 },
    { ...mockGeolocation, id: 2 },
    { ...mockGeolocation, id: 3 },
  ];

  beforeAll(() => {
    const mockGeolocationApiInstance = new MockAdapter(geolocationApiInstance);
    mockGeolocationApiInstance
      .onGet('')
      .reply(200, { results: expectedResults, generationtime_ms: 0 });
  });

  it('snapshot', () => {
    const component = render(
      <AppContext.Provider
        value={{
          geolocation: mockGeolocation,
          setGeolocation: jest.fn(),
          weather: null,
          setWeather: jest.fn(),
        }}
      >
        <SearchGeolocationComponent />
      </AppContext.Provider>,
    );

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('snapshot no data', () => {
    const component = render(
      <AppContext.Provider
        value={{
          geolocation: null,
          setGeolocation: jest.fn(),
          weather: null,
          setWeather: jest.fn(),
        }}
      >
        <SearchGeolocationComponent />
      </AppContext.Provider>,
    );

    expect(component.asFragment()).toMatchSnapshot();
  });

  it('should render the Autocomplete component', async () => {
    render(<SearchGeolocationComponent />);
    const autocomplete = screen.getByTestId('autocomplete');
    expect(autocomplete).toBeInTheDocument();
  });

  it('should render the TextField component', async () => {
    render(<SearchGeolocationComponent />);
    const textField = screen.getByTestId('textField');
    expect(textField).toBeInTheDocument();
  });

  it('should load options when user types in the search field', async () => {
    render(
      <AppContext.Provider
        value={{
          geolocation: null,
          setGeolocation: jest.fn(),
          weather: null,
          setWeather: jest.fn(),
        }}
      >
        <SearchGeolocationComponent />
      </AppContext.Provider>,
    );

    const input = screen.getByRole('combobox');

    await act(() => {
      fireEvent.change(input, { target: { value: mockGeolocation.name } });
    });

    const autocompleteOptions = screen.getAllByTestId('autocomplete-option');

    expect(autocompleteOptions.length).toBe(expectedResults.length);
  });

  it('should set the selected geolocation when the user selects an option', async () => {
    const setGeolocation = jest.fn();

    render(
      <AppContext.Provider
        value={{
          geolocation: null,
          setGeolocation,
          weather: null,
          setWeather: jest.fn(),
        }}
      >
        <SearchGeolocationComponent />
      </AppContext.Provider>,
    );

    const input = screen.getByRole('combobox');

    await act(() => {
      fireEvent.change(input, { target: { value: mockGeolocation.name } });
    });

    const autocompleteOptions = screen.getAllByTestId('autocomplete-option');

    await act(() => {
      fireEvent.click(autocompleteOptions[0]);
    });

    expect(setGeolocation).toBeCalledTimes(1);
  });
});
