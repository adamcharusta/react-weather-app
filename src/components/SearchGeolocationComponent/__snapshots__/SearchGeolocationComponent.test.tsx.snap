// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchGeolocationComponent snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-1b70206-MuiAutocomplete-root"
        data-testid="autocomplete"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          data-testid="textField"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="true"
            for="geolocation-search"
            id="geolocation-search-label"
          >
            Location
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id="geolocation-search"
              role="combobox"
              spellcheck="false"
              type="text"
              value="White House"
            />
            <div
              class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Clear"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
                tabindex="-1"
                title="Clear"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                  data-testid="CloseIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-14lo706"
              >
                <span>
                  Location
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SearchGeolocationComponent snapshot no data 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
    >
      <div
        class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1b70206-MuiAutocomplete-root"
        data-testid="autocomplete"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
          data-testid="textField"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for="geolocation-search"
            id="geolocation-search-label"
          >
            Location
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
              id="geolocation-search"
              role="combobox"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-yjsfm1"
              >
                <span>
                  Location
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
